File: Baekjoon/Class3/40.1931 회의실 배정.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct Meeting {
    int start, end;
};

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n;
    cin >> n;

    vector<Meeting> meetings(n);
    for (int i = 0; i < n; i++)
        cin >> meetings[i].start >> meetings[i].end;
    sort(meetings.begin(), meetings.end(), [](const Meeting& m1, const Meeting& m2) {
        if (m1.end != m2.end) return m1.end < m2.end;
        return m1.start < m2.start;
    }); // 끝나는 시간 기준 정렬

    int max_count = 1; // 첫 미팅
    int prev_end = meetings[0].end;
    for (int i = 1; i < n; i++)
        if (prev_end <= meetings[i].start) { // 다음 미팅 추가
            max_count++;
            prev_end = meetings[i].end;
        }

    cout << max_count;
    return 0;
}

File: Baekjoon/Class3/40.1931 회의실 배정.java
import java.io.*;
import java.util.Arrays;
import java.util.StringTokenizer;

class Main {
    private static class Meeting {
        int start, end;
        public Meeting(int start, int end) { this.start = start; this.end = end; }
    }

    public static void main(String[] args) throws IOException {
        int n;
        Meeting[] meetings;
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        n = Integer.parseInt(br.readLine());
        meetings = new Meeting[n];
        for(int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            meetings[i] = new Meeting(Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken()));
        }
        Arrays.sort(meetings, (m1, m2) -> {
            if(m1.end != m2.end) return m1.end - m2.end;
            return m1.start - m2.start;
        }); // 끝나는 시간 기준 정렬

        int maxCount = 1; // 첫 미팅
        int prevEnd = meetings[0].end;
        for(int i = 1; i < n; i++)
            if(prevEnd <= meetings[i].start) { // 다음 미팅 추가
                maxCount++;
                prevEnd = meetings[i].end;
            }

        System.out.print(maxCount);
        br.close();
    }
}